File: client-sdks/javascript/run_tests.js
#!/usr/bin/env node
/**
 * Test runner for JavaScript SDK with comprehensive testing options
 */

const { spawn } = require("child_process");
const fs = require("fs");
const path = require("path");

function runCommand(command, args = [], options = {}) {
  return new Promise((resolve, reject) => {
    console.log(`Running: ${command} ${args.join(" ")}`);

    const child = spawn(command, args, {
      stdio: "inherit",
      shell: true,
      ...options,
    });

    child.on("close", (code) => {
      if (code === 0) {
        resolve(code);
      } else {
        reject(new Error(`Command failed with exit code ${code}`));
      }
    });

    child.on("error", (error) => {
      reject(error);
    });
  });
}

async function installDependencies() {
  console.log("📦 Installing dependencies...");
  try {
    await runCommand("npm", ["install"]);
    console.log("✅ Dependencies installed successfully");
  } catch (error) {
    console.error("❌ Failed to install dependencies:", error.message);
    process.exit(1);
  }
}

async function runUnitTests() {
  console.log("🧪 Running unit tests...");
  try {
    await runCommand("npm", [
      "test",
      "--",
      "--testNamePattern=^(?!.*Integration)",
    ]);
    console.log("✅ Unit tests completed");
  } catch (error) {
    console.error("❌ Unit tests failed:", error.message);
    process.exit(1);
  }
}

async function runIntegrationTests() {
  console.log("🔗 Running integration tests...");
  try {
    await runCommand("npm", ["test", "--", "--testNamePattern=Integration"]);
    console.log("✅ Integration tests completed");
  } catch (error) {
    console.error("❌ Integration tests failed:", error.message);
    process.exit(1);
  }
}

async function runAllTests() {
  console.log("🚀 Running all tests...");
  try {
    await runCommand("npm", ["test"]);
    console.log("✅ All tests completed");
  } catch (error) {
    console.error("❌ Tests failed:", error.message);
    process.exit(1);
  }
}

async function runCoverage() {
  console.log("📊 Running tests with coverage...");
  try {
    await runCommand("npm", ["test", "--", "--coverage"]);
    console.log("📈 Coverage report generated");
  } catch (error) {
    console.error("❌ Coverage tests failed:", error.message);
    process.exit(1);
  }
}

async function runLint() {
  console.log("🔍 Running linter...");
  try {
    await runCommand("npm", ["run", "lint"]);
    console.log("✅ Linting completed");
  } catch (error) {
    console.error("❌ Linting failed:", error.message);
    process.exit(1);
  }
}

async function testExampleUsage() {
  console.log("💡 Testing example usage...");

  try {
    const { AnomalyClient } = require("./anomaly-client");

    // Test basic initialization
    const client = new AnomalyClient("http://localhost:4000", "test-key");
    console.log("✅ SDK imports and initializes correctly");

    // Test health check if service is running
    try {
      const health = await client.healthCheck();
      console.log("✅ Health check successful:", health);
    } catch (error) {
      console.log(
        "⚠️  Health check failed (service may not be running):",
        error.message
      );
    }

    // Test middleware creation
    const middleware = client.expressMiddleware("test-service");
    if (typeof middleware === "function") {
      console.log("✅ Express middleware created successfully");
    }

    const koaMiddleware = client.koaMiddleware("test-service");
    if (typeof koaMiddleware === "function") {
      console.log("✅ Koa middleware created successfully");
    }

    console.log("✅ Example usage test completed");
  } catch (error) {
    console.error("❌ Example usage test failed:", error.message);
    return false;
  }

  return true;
}

function showHelp() {
  console.log(`
🧪 JavaScript SDK Test Runner

Usage: node run_tests.js [options]

Options:
  --install      Install dependencies
  --unit         Run unit tests only
  --integration  Run integration tests only (requires running data collector)
  --coverage     Run tests with coverage report
  --lint         Run linter
  --example      Test example usage
  --all          Run all tests (default)
  --help         Show this help message

Examples:
  node run_tests.js --install --all     # Install deps and run all tests
  node run_tests.js --unit              # Run only unit tests
  node run_tests.js --coverage          # Run with coverage
  node run_tests.js --example           # Test basic functionality
`);
}

async function main() {
  const args = process.argv.slice(2);

  // Change to script directory
  process.chdir(__dirname);

  if (args.includes("--help")) {
    showHelp();
    return;
  }

  if (args.includes("--install")) {
    await installDependencies();
  }

  if (args.includes("--lint")) {
    await runLint();
  }

  if (args.includes("--unit")) {
    await runUnitTests();
  } else if (args.includes("--integration")) {
    await runIntegrationTests();
  } else if (args.includes("--coverage")) {
    await runCoverage();
  } else if (args.includes("--example")) {
    await testExampleUsage();
  } else if (args.includes("--all") || args.length === 0) {
    await runAllTests();
  }
}

if (require.main === module) {
  main().catch((error) => {
    console.error("❌ Test runner failed:", error.message);
    process.exit(1);
  });
}

module.exports = {
  installDependencies,
  runUnitTests,
  runIntegrationTests,
  runAllTests,
  runCoverage,
  testExampleUsage,
};

File: client-sdks/python/run_tests.py
#!/usr/bin/env python3
"""
Test runner for Python SDK with comprehensive testing options
"""

import subprocess
import sys
import argparse
import os


def run_command(cmd, check=True):
    """Run a command and return the result"""
    print(f"Running: {' '.join(cmd)}")
    result = subprocess.run(cmd, check=check, capture_output=True, text=True)
    print(result.stdout)
    if result.stderr:
        print(result.stderr, file=sys.stderr)
    return result


def install_dependencies():
    """Install test dependencies"""
    print("📦 Installing test dependencies...")
    run_command([sys.executable, "-m", "pip", "install", "-r", "requirements.txt"])


def run_unit_tests():
    """Run unit tests only"""
    print("🧪 Running unit tests...")
    run_command(
        [
            sys.executable,
            "-m",
            "pytest",
            "test_anomaly_client.py",
            "-v",
            "--tb=short",
            "-m",
            "not integration",
        ]
    )


def run_integration_tests():
    """Run integration tests (requires running data collector)"""
    print("🔗 Running integration tests...")
    run_command(
        [
            sys.executable,
            "-m",
            "pytest",
            "test_anomaly_client.py",
            "-v",
            "--tb=short",
            "-m",
            "integration",
        ]
    )


def run_all_tests():
    """Run all tests"""
    print("🚀 Running all tests...")
    run_command(
        [sys.executable, "-m", "pytest", "test_anomaly_client.py", "-v", "--tb=short"]
    )


def run_coverage():
    """Run tests with coverage report"""
    print("📊 Running tests with coverage...")
    run_command(
        [
            sys.executable,
            "-m",
            "coverage",
            "run",
            "-m",
            "pytest",
            "test_anomaly_client.py",
            "-v",
        ]
    )
    run_command([sys.executable, "-m", "coverage", "report"])
    run_command([sys.executable, "-m", "coverage", "html"])
    print("📈 Coverage report generated in htmlcov/")


def test_example_usage():
    """Test the example usage in the SDK"""
    print("💡 Testing example usage...")
    try:
        # Try to import and create client
        from anomaly_client import AnomalyClient

        client = AnomalyClient("http://localhost:4000", api_key="test-key")
        print("✅ SDK imports and initializes correctly")

        # Test health check if service is running
        try:
            health = client.health_check()
            print(f"✅ Health check successful: {health}")
        except Exception as e:
            print(f"⚠️  Health check failed (service may not be running): {e}")

    except ImportError as e:
        print(f"❌ Import error: {e}")
        return False

    return True


def main():
    parser = argparse.ArgumentParser(description="Test runner for Python SDK")
    parser.add_argument("--install", action="store_true", help="Install dependencies")
    parser.add_argument("--unit", action="store_true", help="Run unit tests only")
    parser.add_argument(
        "--integration", action="store_true", help="Run integration tests only"
    )
    parser.add_argument(
        "--coverage", action="store_true", help="Run with coverage report"
    )
    parser.add_argument("--example", action="store_true", help="Test example usage")
    parser.add_argument("--all", action="store_true", help="Run all tests (default)")

    args = parser.parse_args()

    # Change to script directory
    os.chdir(os.path.dirname(os.path.abspath(__file__)))

    if args.install:
        install_dependencies()

    if args.unit:
        run_unit_tests()
    elif args.integration:
        run_integration_tests()
    elif args.coverage:
        run_coverage()
    elif args.example:
        test_example_usage()
    elif args.all or not any(
        [args.unit, args.integration, args.coverage, args.example]
    ):
        run_all_tests()


if __name__ == "__main__":
    main()
